Primitive Jit Logging

Krovatkin opened the pull request.

Krovatkin committed:
jit logging

undo changes to graph_executor.cpp

tweak JIT_GRAPH_DUMP

address Mikhail's feedback

add file filter

break in prod build

address Zach's feedback

address Zach's comments

remove an extra newline

remove GRAPH_DUMP from dce for now

revert all changes in dce

revert all changes to dce

remove unnecessary headers

ZolotukhinM commented on torch/csrc/jit/jit_log.h:
```diff
@@ -0,0 +1,37 @@
+    std::cerr << add_jit_log_prefix(level, ::c10::str(__VA_ARGS__));          \
+  }
+
+// print graph after every pass
```
Nit: The comment is innacurate - it prints graph whenever it is called, not after every pass (even though the intended use might be to print after each pass).

Krovatkin replied:
😄 yes, will fix the comment

ZolotukhinM commented on torch/csrc/jit/jit_log.h:
```diff
@@ -0,0 +1,37 @@
+
+enum class JitLoggingLevels {
+  OFF,
+  GRAPH_DUMP,
```
Nit: I'm not sure these levels would actually be convenient to use(i.e. would I always want to see graph dumps if I'm interested in `info`?), but it works for now :)

Krovatkin replied:
I think, we can figure out which levels we need and what those levels should log as we go. For now, I was thinking to use `INFO` to report individual transformations (we removed a node, constant folded an expression). It usually makes sense to also have graphs after those transformations handy. You could also filter out `JIT_GRAPH_DUMP` if you only need `INFO` messages without graph dumps. It's not ideal, but it's very simple, though.

zdevito replied:
I am not a fan of options. They are not discoverable, people will mix them in poor ways. I feel like we should either always report the thread ID or never report it. Honestly never is better since the compiler almost always is running in a single thread now.

ZolotukhinM commented on torch/csrc/jit/jit_log.cpp:
```diff
@@ -0,0 +1,63 @@
+static std::string level_and_tid(JitLoggingLevels level) {
+  std::stringstream ss;
+  ss << "[" << level << " T";
+  ss << std::this_thread::get_id();
```
Is it always useful? It kind of contaminates the output, so I'd make it optional unless it's really useful.

Krovatkin replied:
that's a great idea. I'll make it optional. Thread ids is very useful when there are multiple concurrent complations which might not be the most common scenario. Log level is also useful when you just want to filter down to the messages at a particular log level.

ZolotukhinM reviewed (approved):
Looks simple and useful!

zdevito commented on torch/csrc/jit/jit_log.cpp:
```diff
@@ -0,0 +1,98 @@
+    const auto str_regexes = caffe2::split(':', c_file_filter);
+
+    for (const auto& str_regex : str_regexes) {
+      file_regexes.emplace_back(std::regex(str_regex));
```
std::regex is seriously broken on platforms like windows. I would not recommend adding a dependency on it.

Krovatkin replied:
I removed this logic since you are against extra options

zdevito reviewed (commented):
My thought here is there is _too much_ stuff going on in the logging that is not discoverable. While you can probably know how to use this, most people won't. I think the following are good constraints:

* There should only be _one_ environment variable that controls logging. It should be obvious how this one environment variable works.
* If you goto definition on the actual JIT_xxx macros, there should be note, within reading distance, about how to use the environment variable.
* Someone who has not read the comments about the logging levels should use be able to use them consistently and correctly. I don't understand DEBUG vs INFO vs GRAPH.
* This isn't JIT specific, having it have a bunch of jit names in it is weird. _especially_ having names like jit _already_ in the jit namespace.

Krovatkin commented:
> There should only be one environment variable that controls logging. It should be obvious how this one environment variable works.

Removed all the other options

> If you goto definition on the actual JIT_xxx macros, there should be note, within reading distance, about how to use the environment variable.

Added a note to the header explaining how to enable logging

> Someone who has not read the comments about the logging levels should use be able to use them consistently and correctly. I don't understand DEBUG vs INFO vs GRAPH.

What would you think would be better names/logging levels? I believe that at least two logging levels would be useful. 
 * The first one reports any transformation made to a graph.
 * The second one should provide enough information to debug a problem in  a particular pass. 



 

But they have very nice comments explaining where to use them.

> This isn't JIT specific, having it have a bunch of jit names in it is weird. especially having names like jit already in the jit namespace.

There's already `C10_LOG`

zdevito commented on torch/csrc/jit/jit_log.h:
```diff
@@ -0,0 +1,41 @@
+// use JIT_GRAPH_DUMP for dumping graphs after optimization passes
+#define JIT_GRAPH_DUMP(MSG, G) \
+  JIT_LOG(JitLoggingLevels::GRAPH_DUMP, MSG, "\n", (G)->toString());
+// use JIT_INFO for reporting graph transformations (i.e. node deletion,
```
INFO vs DEBUG is still confusing to me. The text here makes the difference clear. Can we change INFO to something like GRAPH_UPDATE, CODE_UPDATE, etc.?

zdevito commented on torch/csrc/jit/jit_log.h:
```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include <string>
+
+// To enable logging please set(export) PYTORCH_JIT_LOG_LEVEL to
```
This comment doesn't seem accurate.
`export PYTORCH_JIT_LOG_LEVEL=JIT_GRAPH_DUMP` is that going to work? I see an `atoi` but no mapping from string to level.

zdevito reviewed (approved):
This seems simple enough now. I just have some minor comments about naming.
